src/client_side_request.cc

```diff
@@ -1222,17 +1223,14 @@ ClientRequestContext::clientRedirectDone(const Helper::Reply &reply)
                     new_request->url = tmpUrl;
```

Comment:
```C++
// prevent broken helpers causing too much damage. If old URL == new URL skip the re-write.
if (urlNote != nullptr && strcmp(urlNote, http->uri))
```

@yadij, this comment/discussion is outside this minimal PR scope, but it may be useful to know what can be done with the above URI comparison long-term. As you know, this code is comparing from-helper URI (`urlNote`) and `http->uri` to detect URI changes. ICAP/eCAP adaptation code (e.g., resetRequest() in this PR) is comparing two effectiveRequestUri()s. AFAICT, the two comparison methods ought to yield different results in some corner cases. Do you think this redirector code can eventually be switched to comparing effectiveRequestUri()s instead?

FWIW, this code is already _reporting_ effectiveRequestUri() difference in the debugs() below, not the difference between URIs actually being compared above...

Reply:
IIRC that use of `http->uri` was an optimization to get the URL from way back when `HttpRequest::uri` has issues with being partial vs absolute vs canonical. Correct code today would always do the `tmpUrl.parse(urlNote);` and only do the performance optimizing skip `if (tmpUrl != old_request->uri)`

As the comment implies this whole condition is (almost) purely performance fix for badly coded helpers. We are just trying to avoid the CPU cycles it takes to clone everything (with possible associated state loss from bugs).

PS. Long term I hope that `http->uri` can disappear completely - we should always be referring to whatever `HttpRequest` is "active" (virgin vs adapted vs redirected).